### DISCUSSION TITLE
Prisma client 3.x -> 4.x, no longer coercing empty relationship into an empty list

### DISCUSSION THREAD

**Discussion by cambridgemike** [prisma/prisma#24731] [Q&A]
### Question

I am upgrading from prisma client 3.x to 4.x. Previously, a 1-n relationship would always exist on the returned object, in the default case it would be an empty array. After upgrading, it seems the default case is now null.  

### How to reproduce (optional)

```js
  const experimentWatchers = await prismaClient.experiment
    .findUnique({
      where: {
        id: experimentId,
      },
    })
    .experimentWatchers({
      where: {
        username: user.user,
      },
    });
```
in 3.x

I could then do `experimentWatchers.length > 0;`

in 4.x experimentWatchers can now be null.

Was this an intentional change (I don't see this in the upgrade notes). Is there a way to return the previous behavior?

### Information about Prisma Schema, Client Queries and Environment (optional)
Schema:

```prisma
model Experiment {
  id                          Int                                  @id @default(autoincrement())
  experimentWatchers            ExperimentWatcher[]
}

model ExperimentWatcher {
  id           Int        @id @default(autoincrement())
  experiment   Experiment @relation(fields: [experimentId], references: [id])
  experimentId Int        @map("experiment_id")
}
```

**Comment by ludralph** [comment]
Hi @cambridgemike 👋 

>  Previously, a 1-n relationship would always exist on the returned object, in the default case it would be an empty array. After upgrading, it seems the default case is now null.

Was the query a `findUnique` query or a `findMany` query? For `findMany`, an empty array is returned but for `findUnique`, null is returned. https://www.prisma.io/docs/orm/reference/prisma-client-reference#model-queries

**Comment by cambridgemike** [comment]
I was able to work around this issue, but to demonstrate the issue, we used to have this function which always returned an array:

```js
export async function getExperimentVersionsForExperiment(
  experimentId: number
): Promise<ExperimentVersion[]> {
  return await prismaClient.experiment
    .findUnique({
      where: {
        id: experimentId,
      },
    })
    .experimentVersions()
}
```

After upgrading prisma to 4.12 typescript would complain that the result could be null `Type 'null' is not assignable to type`.

To fix I updated the function to 

```js
import { ExperimentVersion } from '@prisma/client';
import { prismaClient } from '../client';

export async function getExperimentVersionsForExperiment(
  experimentId: number
): Promise<ExperimentVersion[]> {
  return await prismaClient.experiment
    .findUnique({
      where: {
        id: experimentId,
      },
    })
    .experimentVersions()
    .then((res) => (res ??= []));
}
```

Is there any way to change the schema or client initialization so that i don't need to make this change everywhere in my app?

**Reply by cambridgemike** [reply]
I had a similar issue with a findMany query, which was harder to fix:

```js
export async function getGraphQLExperiments(findMany?: FindManyArg): Promise<_GraphQLExperiment[]> {
  return await prismaClient.experiment.findMany({
    ...findMany,
    include: {
      experimentVersions: {
        include: {
          experimentGroup: true,
        },
      },
      experimentUsers: true,
    },
  }).then((result) => {
    return result.map(convertToGraphQLExperiment);
  })
}

function convertToGraphQLExperiment(exp: _GraphQLExperimentPartial): _GraphQLExperiment {
  exp.experimentUsers ??= [];
  exp.experimentVersions ??= [];
  return exp;
}
```

**Reply by cambridgemike** [reply]
I tried using `Prisma.ExperimentGetPayload<{ include: {...` but it did not work as expected, it seems that it keeps the relation types as ? optional

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
